### PR TITLE
[KYUUBI #3935] Support use Trino client to submit SQL

### LINKED ISSUE #3935: [Subtask] Compatiable with Trino query id, slug and token
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues) and found no similar issues.

### Describe the subtask

TODO

### Parent issue

#3901

### Are you willing to submit PR?

- [x] Yes. I can submit a PR independently to improve.
- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by iodone**
### _Why are the changes needed?_

Close #3935 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by iodone** [comment]
cc @ulysses-you

**Comment by ulysses-you** [comment]
thank you @iodone , does this pr base

 on https://github.com/apache/kyuubi/pull/4182 ? I see some duplicate code.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala [hunk 1]
```diff
@@ -91,7 +91,7 @@ trait BackendService {
       foreignTable: String): OperationHandle
   def getQueryId(operationHandle: OperationHandle): String
 
-  def getOperationStatus(operationHandle: OperationHandle): OperationStatus
+  def getOperationStatus(operationHandle: OperationHandle, maxWait: Long = 0): OperationStatus
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala:94]: can we preverse the old method ? we can have both of these:
```
def getOperationStatus(operationHandle: OperationHandle): OperationStatus
def getOperationStatus(operationHandle: OperationHandle, waitTimeMs: Long): OperationStatus
```

Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala:94]: can we preserve the old method ? we can have both of these:
```
def getOperationStatus(operationHandle: OperationHandle): OperationStatus
def getOperationStatus(operationHandle: OperationHandle, waitTimeMs: Long): OperationStatus
```

File: kyuubi-server/pom.xml [hunk 2]
```diff
@@ -221,6 +221,16 @@
...
         </dependency>
 
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jdk8</artifactId>
```
Comment (ulysses-you) [inline kyuubi-server/pom.xml:231]: does it compatiable with JDK11 and 17 ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+      context: TrinoContext,
+      backendService: BackendService): SessionHandle = {
+    context.session
+      .get("sessionId")
+      .fold(backendService.openSession(
+        TProtocolVersion.HIVE_CLI_SERVICE_PROTOCOL_V11,
+        context.user,
+        "",
+        context.remoteUserAddress.getOrElse(""),
+        context.session))(SessionHandle.fromUUID)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala:165]: what does this do ? Trino does not have `session` concept, we can just open a new session for next query with its conext.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/QueryManager.scala [hunk 4]
```diff
@@ -17,13 +17,15 @@
 
 package org.apache.kyuubi.server.trino.api
 
-import javax.ws.rs.ext.ContextResolver
+import java.util.concurrent.ConcurrentHashMap
 
-import com.fasterxml.jackson.databind.ObjectMapper
-import com.fasterxml.jackson.module.scala.DefaultScalaModule
+class QueryManager() {
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/QueryManager.scala:22]: is this code been used ? seems no caller side

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/v1/StatementResource.scala [hunk 5]
```diff
@@ -50,11 +59,32 @@ private[v1] class StatementResource extends ApiRequestContext with Logging {
...
   @Path("/")
   @Consumes(Array(MediaType.TEXT_PLAIN))
-  def query(statement: String, @Context headers: HttpHeaders): QueryResults = {
-    throw new UnsupportedOperationException
+  def query(
+      statement: String,
+      @Context headers: HttpHeaders,
+      @Context uriInfo: UriInfo): Response = {
+    if (statement == null || statement.isEmpty) {
+      throw badRequest(BAD_REQUEST, "SQL statement is empty")
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/v1/StatementResource.scala:70]: badRequest has aleady throwed. how about renaming it as `throwBadRequest` ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/v1/StatementResource.scala [hunk 6]
```diff
@@ -50,11 +59,32 @@ private[v1] class StatementResource extends ApiRequestContext with Logging {
...
+
+    try {
+      val query = Query(statement, trinoContext, translator, fe.be)
+      val qr = query.getQueryResults(query.getLastToken, uriInfo)
+      TrinoContext.buildTrinoResponse(qr, query.context)
+    } catch {
+      case e: Exception =>
+        val errorMsg =
+          s"Error submitting sql"
+        e.printStackTrace()
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/v1/StatementResource.scala:84]: unnecessary

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+  private val slug: Slug = Slug.createNewWithUUID(queryId.getQueryId)
+  private val lastToken = new AtomicLong
+
+  private val defaultMaxRows = 1000
+  private val defaultFetchOrientation = FetchOrientation.withName("FETCH_NEXT")
+
+  def getQueryResults(token: Long, uriInfo: UriInfo, maxWait: Long = 0): QueryResults = {
+    val status =
+      be.getOperationStatus(queryId.operationHandle, maxWait)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala:54]: inline

File: pom.xml [hunk 8]
```diff
@@ -182,6 +182,7 @@
         <scopt.version>4.1.0</scopt.version>
         <slf4j.version>1.7.36</slf4j.version>
         <snakeyaml.version>1.33</snakeyaml.version>
+        <jackson.datatype.version>2.12.3</jackson.datatype.version>
```
Comment (ulysses-you) [inline pom.xml:185]: can we move this close to  <jackson.version> ? and why does these version different ?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/trino/api/v1/StatementResourceSuite.scala [hunk 9]
```diff
@@ -17,20 +17,77 @@
...
-    FrontendProtocols.TRINO :: Nil
+class StatementResourceSuite extends TrinoRestFrontendTestHelper {
 
   test("statement test") {
     val response = webTarget.path("v1/statement/test").request().get()
     val result = response.readEntity(classOf[Ok])
     assert(result == new Ok("trino server is running"))
   }
 
+  test("statement submit and get result") {
```
Comment (ulysses-you) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/trino/api/v1/StatementResourceSuite.scala:42]: can we add negative test for get result ?

**Comment by ulysses-you** [comment]
@iodone  please fix conflicts, thank you

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala (hunk 1 above)
Comment (iodone) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala:94]: `Backendservice` mixed with `BackendServiceMetric`, using multiple inheritance to implement the metric getting wrapper, must have the same parent class. Adding a method will cause a `BackenServiceMetric` compilation error

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala (hunk 3 above)
Comment (iodone) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala:165]: Yes, this point is also to be further confirmed. Each query corresponds to a new session (per query per session), and then the association between them can be passed through the context. Is this correct?

Comment (iodone) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala:165]: Another question, do we need to actively close the session after the results are fetched?

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala:165]: > Each query corresponds to a new session (per query per session), and then the association between them can be passed through the context

yea, because all context is in http header for each query.

> do we need to actively close the session after the results are fetched

What' the behavior of trino-cli ? I remember Trino jdbc driver will send cancel query request when closing connection, then we do not need to eagerly close session.
Besides, Kyuubi session manager has a timeout checker that can close inactive session. So I think we do not need to close session manually.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala (hunk 1 above)
Comment (ulysses-you) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala:94]: I see. Can we remove the default value 0 ? It may cause potential issue.
`def getOperationStatus(operationHandle: OperationHandle, maxWaitMs: Long): OperationStatus`

File: kyuubi-server/pom.xml (hunk 2 above)
Comment (iodone) [reply kyuubi-server/pom.xml:231]: Yes, the data type we currently declare in the Trino HTTP req/res DTO are of java8 DataType. [Java 8 Datatypes](https://github.com/FasterXML/jackson-modules-java8/blob/2.15/datatypes): support for other new Java 8 datatypes outside of date/time: most notably Optional, OptionalLong, OptionalDouble.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala [hunk 10]
```diff
@@ -91,7 +91,8 @@ trait BackendService {
       foreignTable: String): OperationHandle
   def getQueryId(operationHandle: OperationHandle): String
 
-  def getOperationStatus(operationHandle: OperationHandle): OperationStatus
+  def getOperationStatus(operationHandle: OperationHandle, maxWait: Long): OperationStatus
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/BackendService.scala:94]: after second thought, how about `maxWait: Option[Long] = None` ? then we do not need a new method `getOperationStatusTimeout`

File: pom.xml [hunk 11]
```diff
@@ -2309,20 +2316,9 @@
         <profile>
             <id>spark-master</id>
             <properties>
-                <spark.version>3.5.0-SNAPSHOT</spark.version>
-                <!-- https://github.com/ThreeTen/threeten-extra/issues/226 -->
-                <threeten.version>1.7.0</threeten.version>
+                <spark.version>3.4.0-SNAPSHOT</spark.version>
```
Comment (ulysses-you) [inline pom.xml:2319]: please remove unnecessary change in this pom file

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoScalaObjectMapper.scala [hunk 12]
```diff
@@ -19,11 +19,23 @@ package org.apache.kyuubi.server.trino.api
...
+    .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+//      .disable(MapperFeature.AUTO_DETECT_CREATORS)
+//      .disable(MapperFeature.AUTO_DETECT_FIELDS)
+//      .disable(MapperFeature.AUTO_DETECT_SETTERS)
+//      .disable(MapperFeature.AUTO_DETECT_GETTERS)
+//      .disable(MapperFeature.AUTO_DETECT_IS_GETTERS)
+//      .disable(MapperFeature.USE_GETTERS_AS_SETTERS)
+//      .disable(MapperFeature.CAN_OVERRIDE_ACCESS_MODIFIERS)
+//      .disable(MapperFeature.INFER_PROPERTY_MUTATORS)
+//      .disable(MapperFeature.ALLOW_FINAL_FIELDS_AS_MUTATORS)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoScalaObjectMapper.scala:37]: please remove unused code

**Review by ulysses-you — COMMENTED** [review]
otherwise looks good, thank you @iodone

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master/branch-1.7
